src/inc/quic_platform_winkernel.h

```diff
@@ -469,58 +469,131 @@ _CxPlatEventWaitWithTimeout(
+
+__kernel_entry NTSYSCALLAPI
+NTSTATUS
+NtCreateIoCompletion (
+    _Out_ PHANDLE IoCompletionHandle,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_opt_ POBJECT_ATTRIBUTES ObjectAttributes,
+    _In_opt_ ULONG Count
+    );
+
+__kernel_entry NTSYSCALLAPI
+NTSTATUS
+NtSetIoCompletion (
+    _In_ HANDLE IoCompletionHandle,
+    _In_opt_ PVOID KeyContext,
+    _In_opt_ PVOID ApcContext,
+    _In_ NTSTATUS IoStatus,
+    _In_ ULONG_PTR IoStatusInformation
+    );
+
+__kernel_entry NTSYSCALLAPI
+NTSTATUS
+NtRemoveIoCompletionEx (
+    _In_ HANDLE IoCompletionHandle,
+    _Out_writes_to_(Count, *NumEntriesRemoved) PFILE_IO_COMPLETION_INFORMATION IoCompletionInformation,
+    _In_ ULONG Count,
+    _Out_ PULONG NumEntriesRemoved,
+    _In_opt_ PLARGE_INTEGER Timeout,
+    _In_ BOOLEAN Alertable
+    );
+
 //
 // Event Queue Interfaces
 //
 
-typedef KEVENT CXPLAT_EVENTQ; // Event queue
-typedef void* CXPLAT_CQE;
+typedef HANDLE CXPLAT_EVENTQ;
+typedef FILE_IO_COMPLETION_INFORMATION CXPLAT_CQE;
 
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+(CXPLAT_EVENT_COMPLETION)(
+    _In_ CXPLAT_CQE* Cqe
+    );
+typedef CXPLAT_EVENT_COMPLETION *CXPLAT_EVENT_COMPLETION_HANDLER;
+
+typedef struct CXPLAT_SQE {
+    CXPLAT_EVENT_COMPLETION_HANDLER Completion;
+} CXPLAT_SQE;
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
 inline
 BOOLEAN
 CxPlatEventQInitialize(
     _Out_ CXPLAT_EVENTQ* queue
     )
 {
-    KeInitializeEvent(queue, SynchronizationEvent, FALSE);
-    return TRUE;
+    return NT_SUCCESS(NtCreateIoCompletion(queue, IO_COMPLETION_ALL_ACCESS, NULL, 0));
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
 inline
 void
 CxPlatEventQCleanup(
     _In_ CXPLAT_EVENTQ* queue
     )
 {
-    UNREFERENCED_PARAMETER(queue);
+    NtClose(*queue);
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
 inline
 BOOLEAN
-_CxPlatEventQEnqueue(
-    _In_ CXPLAT_EVENTQ* queue
+CxPlatEventQEnqueue(
+    _In_ CXPLAT_EVENTQ* queue,
+    _In_ CXPLAT_SQE* sqe
     )
 {
-    KeSetEvent(queue, IO_NO_INCREMENT, FALSE);
-    return TRUE;
+    return NT_SUCCESS(NtSetIoCompletion(*queue, NULL, sqe, STATUS_SUCCESS, 0));
 }
 
-#define CxPlatEventQEnqueue(queue, sqe) _CxPlatEventQEnqueue(queue)
-
+_IRQL_requires_max_(PASSIVE_LEVEL)
 inline
 uint32_t
 CxPlatEventQDequeue(
     _In_ CXPLAT_EVENTQ* queue,
-    _Out_ CXPLAT_CQE* events,
+    _Out_writes_to_(count, return) CXPLAT_CQE* events,
     _In_ uint32_t count,
     _In_ uint32_t wait_time // milliseconds
     )
 {
-    UNREFERENCED_PARAMETER(count);
-    *events = NULL;
-    return STATUS_SUCCESS == _CxPlatEventWaitWithTimeout(queue, wait_time) ? 1 : 0;
+    LARGE_INTEGER timeout;
+    timeout.QuadPart = -10000LL * wait_time;
+
+    ULONG entriesRemoved = 0;
+    NTSTATUS status = NtRemoveIoCompletionEx(
+        *queue,
+        events,
+        count,
+        &entriesRemoved,
+        (wait_time == UINT32_MAX) ? NULL : &timeout,
+        FALSE
+    );
+
+    if (!NT_SUCCESS(status)) {
+        return 0;
+    }
+    return entriesRemoved;
```

Comment:
```suggestion
    NTSTATUS status;
    ULONG entriesRemoved;
    if (wait_time == UINT32_MAX) {
        status =
            NtRemoveIoCompletionEx(
                *queue,
                events,
                count,
                &entriesRemoved,
                NULL,
                FALSE);
    } else {
        LARGE_INTEGER timeout;
        timeout.QuadPart = -10000LL * wait_time;
        status =
            NtRemoveIoCompletionEx(
                *queue,
                events,
                count,
                &entriesRemoved,
                &timeout,
                FALSE);
    }
    if (!NT_SUCCESS(status)) {
        return 0;
    }
    return entriesRemoved;
```
If we do it this way, then we don't have to calculate the `timeout` if there is no `wait_time`.